lib/segment/src/types.rs

```diff
@@ -505,51 +502,140 @@ pub enum PayloadStorageType {
 pub struct SegmentConfig {
     pub vector_data: HashMap<String, VectorDataConfig>,
-    /// Type of index used for search
-    pub index: Indexes,
-    /// Type of vector storage
-    pub storage_type: StorageType,
+    /// Whether this segment should support appending new data.
+    pub appendable: bool,
     /// Defines payload storage type
-    #[serde(default)]
     pub payload_storage_type: PayloadStorageType,
-    /// Quantization parameters. If none - quantization is disabled.
-    #[serde(default)]
-    pub quantization_config: Option<QuantizationConfig>,
 }
 
 impl SegmentConfig {
     /// Helper to get vector specific quantization config.
     ///
-    /// This grabs the quantization config for the given vector name if it exists. Falls back to
-    /// the collection quantization config.
+    /// This grabs the quantization config for the given vector name if it exists.
     ///
     /// If no quantization is configured, `None` is returned.
     pub fn quantization_config(&self, vector_name: &str) -> Option<&QuantizationConfig> {
         self.vector_data
             .get(vector_name)
             .and_then(|v| v.quantization_config.as_ref())
-            .or(self.quantization_config.as_ref())
     }
 
-    pub fn is_vector_indexed(&self) -> bool {
-        match self.index {
-            Indexes::Plain {} => false,
-            Indexes::Hnsw(_) => true,
-        }
+    pub fn is_any_vector_indexed(&self) -> bool {
+        self.vector_data
+            .values()
+            .any(|config| config.index.is_indexed())
     }
 
-    pub fn is_memmaped(&self) -> bool {
-        match self.storage_type {
-            StorageType::InMemory => false,
-            StorageType::Mmap => true,
-        }
+    pub fn are_all_vectors_indexed(&self) -> bool {
+        self.vector_data
+            .values()
+            .all(|config| config.index.is_indexed())
+    }
+
+    /// Check if any vector storage is mmap'ed
+    ///
+    /// This assumes that when the storage for a named vector requests to be stored on disk; it is
+    /// mmap'ed.
+    pub fn is_any_mmap(&self) -> bool {
+        self.vector_data.values().any(|config| config.on_disk)
     }
 }
 
 /// Config of single vector data storage
 #[derive(Debug, Deserialize, Serialize, JsonSchema, Clone)]
 #[serde(rename_all = "snake_case")]
 pub struct VectorDataConfig {
+    /// Size/dimensionality of the vectors used
+    pub size: usize,
+    /// Type of distance function used for measuring distance between vectors
+    pub distance: Distance,
+    /// Type of index used for search
+    pub index: Indexes,
+    /// Vector specific quantization config that overrides collection config
+    pub quantization_config: Option<QuantizationConfig>,
+    /// False to keep in memory, true to store on disk.
+    /// If true, memory mapped files are used. Then search performance is defined by disk speed and
+    /// the fraction fraction of vectors that fit in RAM.
+    pub on_disk: bool,
```

Comment:
Vector level non-optional `on_disk` flag?

For example, used to decide between in-memory storage and mmap storage.

Replaces `StorageType`, and allows for more options in conjunction with the `appendable` flag.